File: Arrays/88-Merge-Sorted-Array/Python/88-Merge-Sorted-Array-Optimal.py
"""
Problem: 88. Merge Sorted Array
Approach: Optimal - Three Pointers from the End

Use three pointers:
i -> last actual element of nums1
j -> last element of nums2
k -> last position of nums1

Place the larger element at position k and move backwards.

Time Complexity: O(m + n)
Space Complexity: O(1)
"""


class Solution:
    def merge(self, nums1: list[int], m: int,
              nums2: list[int], n: int) -> None:

        i = m - 1
        j = n - 1
        k = m + n - 1

        while j >= 0:

            if i >= 0 and nums1[i] > nums2[j]:
                nums1[k] = nums1[i]
                i -= 1

            else:
                nums1[k] = nums2[j]
                j -= 1

            k -= 1

File: Arrays/88-Merge-Sorted-Array/JavaScript/88-Merge-Sorted-Array-Optimal.js
/**
 * Problem: 88. Merge Sorted Array
 * Approach: Optimal - Three Pointers from the End
 *
 * Use three pointers:
 * i -> last actual element of nums1
 * j -> last element of nums2
 * k -> last position of nums1
 *
 * Place the larger element at position k and move backwards.
 *
 * Time Complexity: O(m + n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void}
 */

var merge = function(nums1, m, nums2, n) {
    let i = m - 1;
    let j = n - 1;
    let k = m + n - 1;

    while (j >= 0) {

        if (i >= 0 && nums1[i] > nums2[j]) {
            nums1[k] = nums1[i];
            i--;
        } else {
            nums1[k] = nums2[j];
            j--;
        }

        k--;
    }
};